signer/sign_state.go

```diff
@@ -577,6 +613,152 @@ func newDiffBlockIDsError(first, second []byte) *DiffBlockIDsError {
 	}
 }
 
+// ConsensusLockViolationError represents an error when trying to sign a block that violates a consensus lock
+type ConsensusLockViolationError struct {
+	msg string
+}
+
+func (e *ConsensusLockViolationError) Error() string { return e.msg }
+
+func newConsensusLockViolationError(
+	lockedValue []byte, requestedValue []byte, height int64, round int64,
+) *ConsensusLockViolationError {
+	return &ConsensusLockViolationError{
+		msg: fmt.Sprintf("consensus lock violation: locked on value %x at height %d round %d, "+
+			"cannot sign different value %x", lockedValue, height, round, requestedValue),
+	}
+}
+
+// ConsensusLockStepViolationError represents an error when trying to sign a step that violates consensus lock rules
+type ConsensusLockStepViolationError struct {
+	msg string
+}
+
+func (e *ConsensusLockStepViolationError) Error() string { return e.msg }
+
+// IsConsensusLockViolationError checks if the error is a consensus lock violation
+func IsConsensusLockViolationError(err error) bool {
+	_, ok := err.(*ConsensusLockViolationError)
+	return ok
+}
+
+// IsConsensusLockStepViolationError checks if the error is a consensus lock step violation
+func IsConsensusLockStepViolationError(err error) bool {
+	_, ok := err.(*ConsensusLockStepViolationError)
+	return ok
+}
```

Comment:
nit: the canonical way to casting errors is by using `errors.As()`.

Reply:
Done thanks